order by nulls first / last [wip]

yurkobb opened the pull request:
I think I'd try implementing Postgres's `ORDER BY ... NULLS { FIRST | LAST }` if you think it's an appropriate feature in deta. I'm not exactly sure about the best api. My most immediate idea is in the test. Alternatively we might use `#:nulls-first` `#:nulls-last`, but that doesn't allow specifying it dynamically (like with `'asc` /  `'desc`). What do you think?

yurkobb committed:
order by nulls first: proposed test [wip]

Bogdanp commented:
I think this would be a good addition. I'd probably go with `#:nulls-first`, `#:nulls-last` and `'nulls-first` and `'nulls-last` to mirror how `#:asc` and `#:desc` work.

yurkobb committed:
implement nulls first/last, update relevant tests

I see three things that may need to be improvement around this:

1. The case of specifying `(order-by q ([tbl.somefield ,'nulls-last]))` is
   currently a contract error since we always treat the second element
   between [] as the ordering direction ('asc / 'desc). I'm not sure
   if we should be trying to implement some kind of runtime matching
   to allow such use or not.

2. If not, perhaps we can provide a better error message than the
   contract error, since it is about internal data structures rather than
   the surface use. This applies to the
   `(order-by q ([tbl.somefield ,'invalid-direction]))` case too.

3. order-by-pair/c is not a pair anymore, but a three-element
   list. Not sure if we should rename it (and the use of `pair` in the
   context of order-by) to something else... Best I could think of was
   `criterion`... meh. Conceptually it is still a pair of column +
   direction (where the direction has as an extra optional nulls flag). A
   two-field struct for the direction? Am I overthinking this?

yurkobb commented:
Pushed a working attempt. There are some rough edges I have outlined in the [commit message](https://github.com/Bogdanp/deta/pull/49/commits/c4ae2c7e98d3bb0cdd01c1f7e772f3bb57711c57) - what do you think?

Bogdanp commented on deta-lib/query.rkt:
```diff
@@ -420,14 +420,22 @@
+                                            (unquote dir-expr:expr)))
+                            (~optional (~or (~and #:nulls-first nulls-first*)
+                                            (~and #:nulls-last  nulls-last* )
+                                            (unquote nulls-expr:expr)))]
```
How about instead of 

```racket
(~optional ...) 
(~optional ...)
```

, we use 

```racket
(~optional 
 (~seq 
  (~or asc desc ...) 
  (~optional (~or nulls ...)))
```
? That way it will be a syntax error to provide a null direction w/o explicitly providing a regular direction, and it'll make the dynamic case less ambiguous.

yurkobb replied:
Fully agree! Pushed.

Bogdanp commented on deta-lib/private/query.rkt:
```diff
@@ -216,7 +216,7 @@
      (query schema opts (struct-copy ast:select stmt [offset (ast:offset n)]))]))
 
 (define order-by-pair/c
```
How about `order-by-config`?

yurkobb replied:
eh, the tough art of naming things... that's an option, though I'm not sure it's the most crystal one because of the config files / command line flags association :/ I was also thinking along the lines of `column-order[ing]/c`, `column-sort/c`, etc., or things like `order-by-cell/c` (or `order-cell/c`, `ordering-cell/c`), `order{-by,ing}-condition/c`, `column-ordering/c`, or even simply `ordering/c`. Or we may settle with `order-by-config/c` if you like it more.

Bogdanp replied:
`ordering/c` sounds best to me out of all of those so let's go with that.

yurkobb committed:
order by: allow nulls-first/last only after explicit asc/desc

yurkobb committed:
internals: rename order-by-pair to ordering

yurkobb committed:
document #:nulls-first / #:nulls-last

Bogdanp commented:
This is great, thanks! Sorry it took so long to merge. I must've missed the notification when you marked it ready for review.
